src/ray/raylet/worker_pool.h

```diff
@@ -39,6 +40,56 @@ namespace raylet {
 using WorkerCommandMap =
     std::unordered_map<Language, std::vector<std::string>, std::hash<int>>;
 
+/// \struct WorkerCacheKey
+///
+/// Struct used to cache workers, keyed by runtime_env. The dynamic_options and
+/// override_environment_variables will be unified into runtime_env in the future.
+struct WorkerCacheKey {
+  std::vector<std::string> dynamic_options;
+  std::unordered_map<std::string, std::string> override_environment_variables;
+  std::string serialized_runtime_env;
+  mutable std::size_t hash_ = 0;
+
+  bool operator==(const WorkerCacheKey &k) const {
+    return dynamic_options == k.dynamic_options &&
+           override_environment_variables == k.override_environment_variables &&
+           serialized_runtime_env == k.serialized_runtime_env;
+  }
+
+  bool EnvIsEmpty() const {
+    return dynamic_options.size() == 0 && override_environment_variables.size() == 0 &&
+           (serialized_runtime_env == "" || serialized_runtime_env == "{}");
+  }
+
+  std::size_t Hash() const {
+    // Cache the hash value.
+    if (!hash_) {
+      if (EnvIsEmpty()) {
+        // It's useful to have the same predetermined value for both unspecified and empty
+        // runtime envs.
+        hash_ = 0;
+      } else {
+        for (auto &str : dynamic_options) {
+          boost::hash_combine(hash_, str);
+        }
+
+        std::vector<std::pair<std::string, std::string>> elems(
+            override_environment_variables.begin(), override_environment_variables.end());
+        std::sort(elems.begin(), elems.end());
```

Comment:
Can you explain why this sort is needed here?

Reply:
Sure, it's that the hash shouldn't depend on the order of the environment variables--I'll add a comment